sdks/go/pkg/beam/core/runtime/exec/datasource.go

```diff
@@ -200,5 +230,41 @@ func (n *DataSource) Progress() ProgressReportSnapshot {
 	if n == nil {
 		return ProgressReportSnapshot{}
 	}
-	return ProgressReportSnapshot{n.SID.Target.ID, n.SID.Target.Name, atomic.LoadInt64(&n.count)}
+	n.mu.Lock()
+	c := n.count
+	n.mu.Unlock()
+	return ProgressReportSnapshot{n.SID.Target.ID, n.SID.Target.Name, c}
+}
+
+// Split takes a sorted set of potential split points, selects and actuates
+// split on an appropriate split point, and returns the selected split point
+// if successful. Returns an error when unable to split.
+func (n *DataSource) Split(splits []int64, frac float32) (int64, error) {
+	if splits == nil {
+		return 0, fmt.Errorf("failed to split: requested splits were empty")
+	}
+	if n == nil {
+		return 0, fmt.Errorf("failed to split at requested splits: {%v}, DataSource not initialized", splits)
+	}
+	n.mu.Lock()
+	c := n.count
+	// Find the smallest split index that we haven't yet processed.
+	p := math.MaxInt64
+
+	for i := int(0); i < len(splits); i++ {
```

Comment:
Consider replacing the two loops with the following single loop, and then we can also remove p as well.
```suggestion
	for _, s := range splits {
		if s >= n.count {
			n.splitPos = s
			fs := n.splitPos
    			n.mu.Unlock()
			return fs, nil 
		}
	}
	n.mu.Unlock()
	...
```
(I might have missed something, but I think these two are the same, and this one is direct)

Reply:
done